src/Compilers/CSharp/Portable/FlowAnalysis/DefiniteAssignment.cs

```diff
@@ -1142,6 +1188,84 @@ protected virtual void ReportUnassigned(Symbol symbol, SyntaxNode node, int slot
 
             // mark the variable's slot so that we don't complain about the variable again
             _alreadyReported[slot] = true;
+
+            bool doesNotNeedImplicitFieldInitialization()
+                => !_requireOutParamsAssigned
+                || CurrentSymbol is not MethodSymbol { MethodKind: MethodKind.Constructor, ContainingType.TypeKind: TypeKind.Struct };
+
+            ErrorCode adjustErrorForStructThis(Symbol thisParameter, int thisSlot)
+            {
+                if (doesNotNeedImplicitFieldInitialization())
+                {
+                    return ErrorCode.ERR_UseDefViolationThis;
+                }
+
+                foreach (var field in _emptyStructTypeCache.GetStructInstanceFields(thisParameter.ContainingType))
+                {
+                    if (_emptyStructTypeCache.IsEmptyStructType(field.Type))
+                        continue;
+
+                    if (field is TupleErrorFieldSymbol)
+                        continue;
+
+                    int slot = VariableSlot(field, thisSlot);
+                    if (slot == -1 || !State.IsAssigned(slot))
+                    {
+                        Debug.Assert((object)field.ContainingType == CurrentSymbol.EnclosingThisSymbol().Type);
+                        AddImplicitlyInitializedField(field);
+                    }
+                }
+
+                if (!compilation.IsFeatureEnabled(MessageID.IDS_FeatureImplicitInitializationInStructConstructors))
```

Comment:
Should this also be conditioned on having found one field needing a default value in the `foreach` loop above?

Reply:
It would perhaps be reasonable to assert that fields were found above, but in this code path we already know that we're reading the 'this' parameter when its flow state is "not definitely assigned".